src/Pos/Delegation/DB.hs

```diff
@@ -1,101 +1,239 @@
+                                         DBnMapIterator, IterType, runDBnIterator,
+                                         runDBnMapIterator)
+import           Pos.DB.Types           (NodeDBs (_gStateDB))
+import           Pos.Delegation.Helpers (isRevokePsk)
+import           Pos.Delegation.Types   (DlgMemPool)
+import           Pos.Types              (ProxySKHeavy, StakeholderId, addressHash)
+import           Pos.Util.Iterator      (nextItem)
 
 ----------------------------------------------------------------------------
 -- Getters/direct accessors
 ----------------------------------------------------------------------------
 
--- | Retrieves certificate by issuer address (hash of public key) if present.
-getPSKByIssuerAddressHash :: MonadDBPure m => StakeholderId -> m (Maybe ProxySKHeavy)
-getPSKByIssuerAddressHash addrHash = gsGetBi (pskKey addrHash)
+-- | Retrieves certificate by issuer public key or his
+-- address/stakeholder id, if present.
+getPskByIssuer
+    :: MonadDBPure m
+    => Either PublicKey StakeholderId -> m (Maybe ProxySKHeavy)
+getPskByIssuer (either addressHash identity -> issuer) = gsGetBi (pskKey issuer)
+
+-- | Given an issuer, retrieves all certificate chains starting in
+-- issuer. This function performs a series of consequental db reads so
+-- it must be used under the shared lock.
+getPskChain
+    :: MonadDBPure m
+    => Either PublicKey StakeholderId -> m DlgMemPool
+getPskChain = getPskChainInternal HS.empty
+
+-- See doc for 'getPskTree'. This function also stops traversal if
+-- encounters anyone in 'toIgnore' set.
+getPskChainInternal
+    :: MonadDBPure m
+    => HashSet StakeholderId -> Either PublicKey StakeholderId -> m DlgMemPool
+getPskChainInternal toIgnore (either addressHash identity -> issuer) =
+    fmap (view _1) $ flip execStateT (HM.empty, [issuer], HS.empty) trav
+  where
+    trav = use _2 >>= \case
+        []                           -> pass
+        (x:_) | HS.member x toIgnore -> (_2 %= drop 1) >> trav
+        (x:_)                        -> do
+            whenM (uses _3 $ HS.member x) $
+                throwM $ DBMalformed "getPskTree: found a PSK cycle"
+            _2 %= drop 1
+            pskM <- lift $ getPskByIssuer $ Right x
+            whenJust pskM $ \psk -> do
+                let is = pskIssuerPk psk
+                _1 %= HM.insert is psk
+                _3 %= HS.insert (addressHash is)
+            trav
+
+-- | Retrieves certificate forest, where given issuers are trees'
+-- leaves. Executes 'getPskChain' for every issuer and merges. This
+-- function must be used under outside shared lock.
+getPskForest
+    :: (MonadDBPure m)
+    => Either [PublicKey] [StakeholderId] -> m DlgMemPool
+getPskForest (either (fmap addressHash) identity -> issuers) =
+    foldlM foldFoo HM.empty (map Right issuers)
+  where
+    -- Don't revisit branches we retrieved earlier.
+    foldFoo cur = getPskChainInternal (HS.fromList $ map addressHash $ HM.keys cur)
```

Comment:
You're right. This function is not used anyway, so i forgot to fix it I guess. Will just delete.